### PR TITLE
feat(angular-todo): Adds centralized login implementation

### PR DESCRIPTION
**Pull request by juangarmendia05**

### CODE REVIEW

**Comment by juangarmendia05** [comment]
image

image

image

image

image

image

File: angular-todo/src/app/app.component.ts [hunk 1]
```diff
@@ -68,7 +68,7 @@ export class AppComponent implements OnInit {
 
     Config.set({
       clientId: environment.WEB_OAUTH_CLIENT,
-      redirectUri: `${window.location.origin}/callback.html`,
+      redirectUri: `${window.location.origin}/${environment.CENTRALIZED_LOGIN === 'true' ? 'login' : 'callback.html'}`,
```
Comment (ryanbas21) [inline angular-todo/src/app/app.component.ts]: Hmm, I think we can leave the redirectUri as the callback.html, right?

File: angular-todo/src/app/views/login/login.component.ts [hunk 2]
```diff
@@ -17,10 +23,43 @@ import { Component } from '@angular/core';
   selector: 'app-login',
   templateUrl: './login.component.html',
 })
-export class LoginComponent {
+export class LoginComponent implements OnInit{
   isWebAuthn = false;
+  code: string;
+  state: string;
+
+  get isCentrlizedLogin(): boolean {
```
Comment (ryanbas21) [inline angular-todo/src/app/views/login/login.component.ts]: can we spell `isCentralizedLogin` correctly? Is there a reason for this spelling?

**Review by ryanbas21 — COMMENTED** [review]
Looks pretty good

File: angular-todo/src/app/app.component.ts (hunk 1 above)
Comment (juangarmendia05) [reply angular-todo/src/app/app.component.ts]: If you leave it as 'callback.html', after successfully logging in the OAuth 2.0 flow redirects to 'OK'. 

image

File: angular-todo/src/app/views/login/login.component.ts (hunk 2 above)
Comment (juangarmendia05) [reply angular-todo/src/app/views/login/login.component.ts]: no reason, my bad, it got away from me.

Comment (ryanbas21) [reply angular-todo/src/app/views/login/login.component.ts]: Ah yes your right sorry.

**Review by ryanbas21 — APPROVED** [review]
I dont have much to add to this. just flip back the redirecturi

File: angular-todo/src/app/app.component.ts (hunk 1 above)
Comment (cerebrl) [reply angular-todo/src/app/app.component.ts]: Juan is right here. The `callback.html` is only good for background requests to `/authorize`. It's not good for the full redirects as the user gets redirected to that HTML file with no action to take.

File: angular-todo/src/app/views/login/login.component.html [hunk 3]
```diff
@@ -8,25 +8,27 @@
...
 
-<div class="cstm_container_v-centered container-fluid d-flex align-items-center">
-  <div class="w-100">
-    <app-back-home></app-back-home>
-    <div class="card shadow-sm p-5 mb-2 w-100">
-      <div class="cstm_form-icon align-self-center mb-3">
-        <ng-container *ngIf="!isWebAuthn; else fingerPrint">
-          <app-key-icon size="72px"></app-key-icon>
-        </ng-container>
+<ng-container *ngIf="!isCentralizedLogin">
```
Comment (cerebrl) [inline angular-todo/src/app/views/login/login.component.html]: Sorry if I'm just missing this, but where's the `else` for this condition? If it's not centralized login, what renders?

File: angular-todo/src/app/views/login/login.component.ts [hunk 4]
```diff
@@ -17,10 +23,43 @@ import { Component } from '@angular/core';
   selector: 'app-login',
   templateUrl: './login.component.html',
 })
-export class LoginComponent {
+export class LoginComponent implements OnInit{
```
Comment (cerebrl) [inline angular-todo/src/app/views/login/login.component.ts]: Can you add a space between `OnInit` and the following `{`?

File: angular-todo/src/app/views/login/login.component.ts [hunk 5]
```diff
@@ -8,7 +8,13 @@
  * of the MIT license. See the LICENSE file for details.
  */
 
-import { Component } from '@angular/core';
+import { Component, OnInit } from '@angular/core';
+import { environment } from '../../../environments/environment';
+import {
+  TokenManager, UserManager,
```
Comment (cerebrl) [inline angular-todo/src/app/views/login/login.component.ts]: It seems like some of this code could use some formatting. Can you try to format this code using VS Code?

File: angular-todo/src/app/views/login/login.component.ts [hunk 6]
```diff
@@ -17,10 +23,43 @@ import { Component } from '@angular/core';
...
+    private route: ActivatedRoute,
+    public userService: UserService,
+    private router: Router,
+  ) {}
+
+  async ngOnInit(): Promise<void> {
+    this.code = this.route.snapshot.queryParamMap.get('code');
+    this.state = this.route.snapshot.queryParamMap.get('state');
+
+    if(this.isCentralizedLogin){
```
Comment (cerebrl) [inline angular-todo/src/app/views/login/login.component.ts]: Formatting issue here too.

Comment (cerebrl) [reply angular-todo/src/app/views/login/login.component.ts]: Here's another if central login, but I see no `else` do this ... ?

**Review by cerebrl — CHANGES_REQUESTED** [review]
I'm seeing some conditions that don't have a fallback, so if centralized login is false, this app is likely broken. I also see some formatting issues, so can you run prettier on these files?

File: angular-todo/src/app/views/login/login.component.html (hunk 3 above)
Comment (juangarmendia05) [reply angular-todo/src/app/views/login/login.component.html]: hi justin, if it is centralized login it redirects to ForgeRock AM, or another web application to handle authentication. If not centralized renders the form to fill out with username and password

File: angular-todo/src/app/views/login/login.component.ts (hunk 6 above)
Comment (juangarmendia05) [reply angular-todo/src/app/views/login/login.component.ts]: It only comes into action if it's centralized login. If not, it follows the normal flow which is rendering the form and asks for the username and password

File: angular-todo/src/app/app.component.ts (hunk 1 above)
Comment (juangarmendia05) [reply angular-todo/src/app/app.component.ts]: If it is centralized login, after successfully logging in i need to be redirected to '/login', which is why i added the conditional ternary operator.

File: angular-todo/src/app/views/login/login.component.ts (hunk 2 above)
Comment (ryanbas21) [reply angular-todo/src/app/views/login/login.component.ts]: the above comment was meant for the callback comment i made earlier so my bad i posted it here

File: angular-todo/src/app/views/login/login.component.html (hunk 3 above)
Comment (cerebrl) [reply angular-todo/src/app/views/login/login.component.html]: If this means that with Centralized Login, the app renders just a blank page, let's change that do rendering the existing spinner component with a message of "Verifying your session", or something. That way, the app doesn't look broken if there's latency in the network calls.

**Review by cerebrl — APPROVED** [review]
This looks good other than the possible blank page being rendered for Centralized Login. If that's addressed, then this is mergable. Thanks.

**Review by cerebrl — APPROVED** [review]

**Pull request merged**